### ISSUE TITLE
Build from source is utterly, TOTALLY broken

### ISSUE THREAD

**Issue by stugol** [felix-lang/felix#76]
I've been trying for the past hour and a half to build Felix on a 64-bit Debian Jessie system.

There are several things wrong with your build process. Allow me to highlight some of them.

1: Undocumented dependancy on the `ocaml-native-compilers` package (for `ocamldep` and `ocamldep.opt`). Spent ten entertaining minutes figuring that one out.

2: `flx_build_flxg` (in GNUmakefile, line 139) is not located in my system path, therefore IT WILL NOT WORK! I've had to edit the makefile and insert `./build/release/host/bin` before it.

3: `flx_build_rtl` (in GNUmakefile, line 152) is not located in my system path, therefore IT WILL NOT WORK! I've had to edit the makefile and insert `./build/release/host/bin` before it.

4: `flx_build_boot` (in GNUmakefile, line 158) is not located in my system path, therefore IT WILL NOT WORK! I've had to edit the makefile and insert `./build/release/host/bin` before it.

5: `toolchain_gcc_linux.so` cannot be found during the compilation process, because it is located in `./build/release/host/lib/rtl/` and - oddly! - NOT MY SYSTEM LIBRARIES DIRECTORY! This leads to the following error, which I have been unable to fix:

```
Dynamic linkage error
filename: toolchain_gcc_linux.so
operation: LoadLibrary/dlopen
what: Cannot find dll/shared library
GNUmakefile:174: recipe for target 'rtl' failed
make: *** [rtl] Error 253
```

6: Btw, the `tarballs` link on [your site](http://felix-lang.org/share/src/web/download_tarball.html) is broken as well.

Now, your project looks really nice, and I was looking forward to coding with it. But after _fifteen years_ of development, such a slipshod build process really doesn't inspire a huge amount of confidence.

Sure, I could use the prebuilt binary - which `curl`, btw, was unable to download for some reason; I had to use `wget` - but who knows when you'll release a new version? I could be months out of date and there'd be nothing I could do about it.

Guys, please fix your project. It'd be great to be able to use it.

**Comment by skaller** [comment]
> On 27 Dec 2015, at 14:23, stugol <notifications@github.com> wrote:

I strongly recommend you join the mailing list.

> 
> I’ve been trying for the past hour and a half to build Felix on a 64-bit Debian Jessie system

It autobuilds on Travis, which is Ubuntu.

> 
> There are several things wrong with your build process Allow me to highlight some of them
> 
> 1: Undocumented dependancy on the ocaml-native-compilers package (for ocamldep and ocamldepopt) Spent ten entertaining minutes figuring that one out

That’s documented:

http://felix-lang.org/share/src/web/download_source.html



> 
> 2: flx_build_flxg (in GNUmakefile, line 139) is not located in my system path, therefore IT WILL NOT WORK! I’ve had to edit the makefile and insert /build/release/host/bin before it

Yes, that’s a recent change sorry.  You need to do this:

. buildscript/linuxsetup.sh

> 6: Btw, the tarballs link on your site is broken as well

Yes, sorry. The tarballs are no generated by Travis (and Appveyor) and
kept on GitHub.

The website is not up to date. Please understand the build system is quite
complex and is under development continuously. Along with the compiler,
libraries, tools, documentation .. this is a BIG project and under-resourced.

The build system alone took several years to develop and is a product
in its own right. It’s cross platform. This is not a Linux project.
The build system works on Linux, OSX, and WIndows (more or less
unmodified). The build system is also used to bootstrap a SECOND
build system which rebuilds the system.


—
john skaller
skaller@users.sourceforge.net
http://felix-lang.org

**Comment by stugol** [comment]
> I strongly recommend you join the mailing list.

Sure. Just as soon as you tell me how to do that. Because there is ZERO mention of this on either the Github page or your own website. Are we expected to be psychic?

> It autobuilds on Travis, which is Ubuntu.

What is meant by "autobuild"?

It may well build perfectly on this "Travis" of yours, but on Debian - which is what I'm using - it does not.

> That’s documented:
> http://felix-lang.org/share/src/web/download_source.html

Fair point. I must have missed that one.

> Yes, that’s a recent change sorry.  You need to do this:
> . buildscript/linuxsetup.sh

Well, I guess that would've been a helpful thing to know. Shame there was no mention of this fact anywhere.

I'll give it another try, see how far I get.

> Yes, sorry. The tarballs are no generated by Travis (and Appveyor) and kept on GitHub.

Ah, cool. Is their location a secret? Because I don't see them.

> The website is not up to date. Please understand the build system is quite complex and is under development continuously. Along with the compiler, libraries, tools, documentation .. this is a BIG project and under-resourced.
>
> The build system alone took several years to develop and is a product in its own right. It’s cross platform. This is not a Linux project. The build system works on Linux, OSX, and WIndows (more or less unmodified). The build system is also used to bootstrap a SECOND build system which rebuilds the system.

This is all very impressive. But how do you expect anyone to use your project if the documentation is wrong? The normal build process for a git project is to ```./configure``` followed by a ```make```, which doesn't work. Your own website gives build instructions which are completely wrong; and provides links to binaries which don't exist.

Strikes me your project will _continue_ to be under-resourced until more people discover and use it. And people will be unable to do that, because they are required to jump through arbitrary hoops that aren't documented anywhere.

Upon experiencing these problems, there are two things people will do. They will either give up and assume the project is broken - in which case your "customer base" (and therefore also contributors) will remain close to zero - or they will do what I did, which is come on here and post a snide remark.

The language looks really interesting, and I'd like to try it out - and maybe contribute to it, if my skill is sufficient. But you're gonna need to meet me half-way here, and write some goddamn documentation. Sheesh!

So, to recap:
   1. Where do the tarballs live, besides the _single_ release on your releases page?
   2. You said it auto-builds, whatever that means. Does this mean a binary is produced whenever you commit code? If so, where do I find these binaries?
   3. Do I need to do anything else, besides running ```linuxsetup.sh``` to build the project? Aside from ```make``` of course. The documentation mentions running ```make rebuild``` after ```make build``` also: what is the reason for this?
   4. Exactly how out of date is the website? If I follow the tutorials and language guides, will my code work? Or have you changed the syntax and not bothered to update any of the documentation? I'm not a mind-reader.

**Comment by kirbyfan64** [comment]
@stugol

- https://groups.google.com/forum/m/#!forum/felix-language contains sign-up info.
- Auto-builds build every commit.
- The tarballs are on the releases page of GitHub. I don't know where it actually is, but that's kind of GitHub's fault! :/
- The website should be pretty much spot-on syntax-wise.

**Comment by stugol** [comment]
Thank you.

I would still like an answer to question #3 though. Anyone?

**Comment by kirbyfan64** [comment]
@stugol You shouldn't have to run anything else. I think `make rebuild` is slightly faster than `make build`, but I'm not sure...

In reality, you should usually just run `make build`.

**Comment by kirbyfan64** [comment]
Haha, it's actually more obvious that I thought. I missed it for some reason!

On the main GitHub page for this project, there's a link to 'Releases'. That's where the tarballs are! The main page should probably link to it, though.

**Comment by skaller** [comment]
> On 28 Dec 2015, at 09:48, stugol <notifications@github.com> wrote:
> 
> Thank you.
> 
> I would still like an answer to question #3 though. Anyone?

For building:

~/felix>cat buildscript/linuxsetup.sh
export PATH=build/release/host/bin:$PATH
export LD_LIBRARY_PATH=build/release/host/lib/rtl:$LD_LIBRARY_PATH

For installing (in GNUmakefile):

install:
        rm -rf ${INSTALLDIR}
        ${BUILDROOT}/host/bin/flx_cp ${BUILDROOT}/host '(.*)' ${INSTALLDIR}'/host/$${1}'
        ${BUILDROOT}/host/bin/flx_cp ${BUILDROOT}/share '(.*)' ${INSTALLDIR}'/share/$${1}'
        ${BUILDROOT}/host/bin/flx_cp ${BUILDROOT} '(VERSION)' ${INSTALLDIR}'/$${1}'
        ${BUILDROOT}/host/bin/flx_cp ${BUILDROOT}/host/bin '(flx)' ${EXECPREFIX}'/$${1}'
        ${BUILDROOT}/host/bin/flx_cp speed/ '(.*)' ${INSTALLDIR}'/speed/$${1}'

        rm -f ${INSTALLROOT}/felix-latest
        ln -s felix-${VERSION} ${INSTALLROOT}/felix-latest

The method of accessing Felix, installed or not, has a lot of complex options.
In other words its extremely versatile :-) By default the install puts “flx” in
/usr/local/bin and everything else in /usr/local/lib/felix/felix-latest (a symlink
to a specific version).

I myself don’t use this, i just run it from build/release. The default is to look
in $HOME/.felix/config/felix.fpc for configuration information. I personally
copied build.fpc there and just run from the build image (being a dev).
However it is important to actually do an install in case the build image
screws up.

The method of configuring Felix has been changed and is still subject
to change. It is being set up to support more or less arbitrary locations
of the components. So it can work for a simple user or a server based
enterprise install with multiple platforms. Getting all this to work is
non trivial when one has a single laptop to play with.

Felix is also able to simultaneously support multiple targets and
multiple versions. Seriously, there are not a lot of other systems
that can come close to this (the Debian autobuilder springs to mind
as something with different but comparably complex requirements).

I personally run OSX and Windows 10, but don’t have a Linux box.
Remote access to the server running Ubuntu for me is restricted
by poor 3G wiresless reception half way around the world from the server.

—
john skaller
skaller@users.sourceforge.net
http://felix-lang.org

**Comment by skaller** [comment]
> On the main GitHub page for this project, there's a link to 'Releases'. That's where the tarballs are! The main page should probably link to it, though.

the “main page” has been half redesigned and never completed several
times now .. :)

—
john skaller
skaller@users.sourceforge.net
http://felix-lang.org

**Comment by kirbyfan64** [comment]
@skaller :D Last time though, no one liked my colorful home page...

**Comment by stugol** [comment]
> I personally run OSX and Windows 10, but don’t have a Linux box.
Remote access to the server running Ubuntu for me is restricted
by poor 3G wiresless reception half way around the world from the server.

Um...Linux is the easiest of the three platforms to run, surely? Can't you just dual-boot it with Windows? Or get a shitty old netbook or something? Seriously, I'm tripping over old PCs here, and I'd be hard-pressed to find one that _wouldn't_ run Linux.

Alternatively, install VirtualBox on one of your systems and install it in there. It'd help if the system supported VT-x, but even without, it should manage a basic command-line. Running X without VT-x would be insane, obviously.

**Comment by skaller** [comment]
Win10 and OSX10 both work out of the box pretty much. In particular they work with the laptops they came on. Linux is much harder to setup, there's no assurance it will work on a given laptop with all the hardware. I was given my new Mac, its perfect for my usage: fast, solid state, low power consumption. I bought the Windows box because I got a great price running my preferred hardware (ASUS), in theory to build Felix on Windows because this is hard to do without a local box. It builds on Appveyor now but there was no way to get a remote build working without a local box. Also I use Windows for .. er .. games!!

If I had the cash I could think about a Linux server, but felix-lang.org runs Ubuntu and we build on Travis too. And if it builds on OSX it will probably work on Linux. So I get the convenience of a pair of popular consumer boxes for non-development, as well as two development boxes for the "harder to build for" platforms. Linux has some advantages, the package management is an obvious one. But I need to be able to support Windows and Mac anyhow without packages.

Both my boxes are core i7 quad cores with 4G and 8G ram. On my old box builds took up to 4 hours. I don't really need a slow build. Its under 20 minutes on both my new machines.

I'm not running any virtual stuff, not only don't I really see the point, I can't afford the bandwidth to download anything substantial. I only have a flaky 3G connection to the internet and it costs over $10/G which is a lot for me. Slow expensive internet is one of the downsides of living on a yacht :)

**Comment by stugol** [comment]
Poor thing, living on a yacht. My heart bleeds for you.

Anyhow, those machines seem eminently capable of running a virtual machine. There's nothing stopping you installing VirtualBox and, say, Debian within it. Sure, you might want to use someone else's internet connection to download Debian...

But yeah, just an idea. If you don't want to do it, I'm not gonna force you.

**Comment by skaller** [comment]
I live on a yacht because I am too poor to rent a shoebox on land. Sydney is one of the most expensive cities in the world.

**Comment by stugol** [comment]
.....blimey. Over here, yachts are generally inhabited by those rich enough to afford them. It must be very different there.
